### PR TITLE
Faster k-quants on older GPUs

### PR DESCRIPTION
**Pull request by ikawrakow**
The oldest GPU I have access to is a GTX-1660 with 6 GB VRAM. The PR significantly improves token prediction on this GPU for `Q2_K - Q4_K` and gets a small boost for `Q5_K` as well (see table). `Q6_K` was not changed. It would be nice to also have some results for the GTX-1000 series (and/or other older GPUs). The PR does not change performance on a modern GPU (with the RTX-4080 that I have available being used for testing). 

| Quantization | t/s master | t/s PR | speed up |
|--:|--:|--:|--:|
| Q2_K | 22.6 | 27.4 | 1.21 |
| Q3_K_S | 17.5 | 24.3 | 1.39 |
| Q4_K_S | 16.3 | 24.4 | 1.50 |
| Q5_K_S | 15.1 | 16.1 | 1.07 |
| Q6_K | 16.5 | 16.5 | 1.00 |
| Q4_0 | 21.4 | - | - |
| Q5_0 | 15.0 | - | - |

### CODE REVIEW

**Comment by KerfuffleV2** [comment]
I tested on a GTX 1060 6GB:

```plaintext
q2_k
master: eval time = 12884.48 ms /    99 runs   (  130.15 ms per token,     7.68 tokens per second)
pull: eval time = 12838.96 ms /    99 runs   (  129.69 ms per token)

q3_k_s
master: eval time = 10787.75 ms /    99 runs   (  108.97 ms per token,     9.18 tokens per second)
pull: eval time = 10813.03 ms /    99 runs   (  109.22 ms per token)

q4_k_s
master: eval time = 18330.58 ms /    99 runs   (  185.16 ms per token,     5.40 tokens per second)
pull: eval time = 18413.52 ms /    99 runs   (  186.00 ms per token)

q5_k_s:
master: eval time = 18408.78 ms /    99 runs   (  185.95 ms per token,     5.38 tokens per second)
pull: eval time = 15824.68 ms /    99 runs   (  159.85 ms per token)
```

TL;DR: q2_k, q3_k, q4_k - no noticeable difference. q5_k actually sped up a decent amount though!

**Comment by JohannesGaessler** [comment]
My test results:

| GPU      | Test   | Model     | t/s master | t/s PR | t/s PR LLAMA_CUDA_KQUANTS_ITER=1 | Speedup |
|----------|--------|-----------|------------|--------|----------------------------------|---------|
| GTX 1070 | TG 128 | 7b Q2_K   |       8.80 |   8.75 |                            11.11 |    1.26 |
| GTX 1070 | TG 128 | 7b Q3_K_S |       9.15 |   9.05 |                             9.78 |    1.07 |
| GTX 1070 | TG 128 | 7b Q4_K_S |       6.61 |   6.60 |                             8.79 |    1.33 |
| GTX 1070 | TG 128 | 7b Q5_K_S |       6.15 |   6.56 |                             6.56 |    1.07 |
| GTX 1070 | TG 128 | 7b Q4_0   |      10.23 |      - |                                - |       - |
| GTX 1070 | TG 128 | 7b Q4_0   |       7.07 |      - |                                - |       - |

**Comment by KerfuffleV2** [comment]
Wow, `LLAMA_CUDA_KQUANTS_ITER` makes a huge difference on my 1060, at least for q4_k.

```plaintext
Q5_K_S: eval time = 15554.61 ms /    99 runs   (  157.12 ms per token)
Q4_K_S: eval time = 10581.56 ms /    99 runs   (  106.88 ms per token)
Q3_K_S: eval time =  9499.68 ms /    99 runs   (   95.96 ms per token)
Q2_K: eval time =  8531.30 ms /    99 runs   (   86.17 ms per token)
```

Nearly doubled Q4_K speed, and a nice bump for Q2_K for anyone crazy enough to actually try to use it.

I also tested with `LLAMA_CUDA_KQUANTS_ITER=1` on master:

```plaintext
eval time = 16926.22 ms /    99 runs   (  170.97 ms per token,     5.85 tokens per second)
```

It's a little faster, but nothing dramatic, so it's the combination of this PR and that tweak which makes the difference. Sadly, I get the best token generation speed of all by not offloading anything (~84ms/token) but people with old GPUs and worse CPUs will definitely see a lot of benefit from the change.

**Comment by johnson442** [comment]
LLAMA_CUDA_KQUANTS_ITER=1 on master and PR with a 1070ti:
|	|Master|	|PR	|	|
|-------|------|--------|-------|-------|	
|7B	|ms/t	|t/s	|ms/t	|t/s	|
|Q2_K	|73.66	|13.58	|57.3	|17.45	|
|Q3_K_S	|70.62	|14.16	|60.7	|16.47	|
|Q4_K_S	|117.94	|8.48	|73.34	|13.64	|
|Q5_K_S	|122.87	|8.14	|117.29	|8.53	|
|Q6_K	|70.32	|14.22	|70.31	|14.22	|
|	|	|	|	|	|
|13B Q3_K_S (-ngl 42 -t 5)|150.96	|6.62	|118.31	|8.45|
|30B Q3_K_S (-ngl 28 -t 5)|357.3	|2.80	|334.31	|2.99|

I'm off to download some Q4_K models!

**Comment by JohannesGaessler** [comment]
I think the table that I just added in README.md regarding compilation options will need to be adapted but otherwise I think this is good to merge.

**Review by ggerganov — APPROVED** [review]
